From 539a1924c0078daf4913643ebee1282767fc5656 Mon Sep 17 00:00:00 2001
From: Ethan Chen <intervigil@gmail.com>
Date: Tue, 17 Nov 2015 14:08:43 -0800
Subject: [PATCH 1/2] Don't attempt to convert degree to orientation enum twice

* Exif.getOrientation already calls getRotationForOrientationValues,
  which returns a degree value. Don't attempt to convert that degree
  value back into degrees, that makes no sense.

Change-Id: Ie35a0683e2d36801a9130ec19e1b3aa1f48b92c2
---
 .../camera/processing/imagebackend/TaskCompressImageToJpeg.java        | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/src/com/android/camera/processing/imagebackend/TaskCompressImageToJpeg.java b/src/com/android/camera/processing/imagebackend/TaskCompressImageToJpeg.java
index 2e5976c..e1d1689 100644
--- a/src/com/android/camera/processing/imagebackend/TaskCompressImageToJpeg.java
+++ b/src/com/android/camera/processing/imagebackend/TaskCompressImageToJpeg.java
@@ -95,8 +95,7 @@ public class TaskCompressImageToJpeg extends TaskJpegEncode {
      */
     public Map<Integer, Integer> exifGetMinimalTags(ExifInterface exif) {
         Map<Integer, Integer> map = new HashMap<>();
-        map.put(ExifInterface.TAG_ORIENTATION,
-                ExifInterface.getRotationForOrientationValue((short) Exif.getOrientation(exif)));
+        map.put(ExifInterface.TAG_ORIENTATION, Exif.getOrientation(exif));
         map.put(ExifInterface.TAG_PIXEL_X_DIMENSION, exif.getTagIntValue(
                 ExifInterface.TAG_PIXEL_X_DIMENSION));
         map.put(ExifInterface.TAG_PIXEL_Y_DIMENSION, exif.getTagIntValue(
-- 
2.7.4


From 7b5c75a2636959f2868dd41de8fb89af68da0724 Mon Sep 17 00:00:00 2001
From: Luden <luden@ghostmail.com>
Date: Fri, 19 Feb 2016 23:19:13 +0100
Subject: [PATCH 2/2] Fixed Camera2 crash on first camera start.

When camera enters video preview mode, it uses current window
size to select the best matching video preview size to request
from the camera HAL.

However, if video preview mode is the initial mode in which the
camera starts, there's a race condition between preview size
selection and camera app layout completion. If preview size
selection code is run first, it will get zero window size and
subsequently will crash.

It's not clear what's really the "best" fix in this situation
(= waiting for layout to complete would introduce extra latency
in the camera startup process), so current work-around just checks
whether window size is already defined and if not - uses display
size instead. This makes sense because normally camera window
is running full-screen anyway, so window size will match display
size.

Change-Id: Ie8555025902ba8ae6f719d45371145126290bb7c
---
 src/com/android/camera/VideoUI.java | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/src/com/android/camera/VideoUI.java b/src/com/android/camera/VideoUI.java
index 9d00969..458bd30 100644
--- a/src/com/android/camera/VideoUI.java
+++ b/src/com/android/camera/VideoUI.java
@@ -19,6 +19,7 @@ package com.android.camera;
 import android.graphics.Bitmap;
 import android.graphics.Point;
 import android.graphics.SurfaceTexture;
+import android.util.DisplayMetrics;
 import android.view.GestureDetector;
 import android.view.MotionEvent;
 import android.view.View;
@@ -246,7 +247,17 @@ public class VideoUI implements PreviewStatusListener {
      * @return The size of the available preview area.
      */
     public Point getPreviewScreenSize() {
-        return new Point(mRootView.getMeasuredWidth(), mRootView.getMeasuredHeight());
+        if (mRootView.getMeasuredWidth() > 0 && mRootView.getMeasuredHeight() > 0) {
+            return new Point(mRootView.getMeasuredWidth(), mRootView.getMeasuredHeight());
+	} else {
+	    // Measured width and height are not yet available - use screen size
+	    // as a proxy, should be fine given camera typically is full screen anyhow.
+	    DisplayMetrics metrics = new DisplayMetrics();
+	    mActivity.getWindowManager().getDefaultDisplay().getMetrics(metrics);
+	    Log.i(TAG, "Measured width and height are not yet available, reverting " +
+	            "to display size: " + metrics.widthPixels + ", " + metrics.heightPixels);
+	    return new Point(metrics.widthPixels, metrics.heightPixels);
+	}
     }
 
     /**
-- 
2.7.4

